src/api_routes_websocket.rs

```diff
@@ -117,131 +131,207 @@ use std::{
+    hb(ctx);
+
+    // register self in chat server. `AsyncContext::wait` register
+    // future within context, but context waits until this future resolves
+    // before processing any other events.
+    // HttpContext::state() is instance of WsChatSessionState, state is shared
+    // across all routes within application
+    let addr = ctx.address();
+    self
+      .apub_data
+      .chat_server()
+      .send(Connect {
+        addr: addr.recipient(),
+      })
+      .into_actor(self)
+      .then(|res, act, ctx| {
+        match res {
+          Ok(res) => act.id = res,
+          // something is wrong with chat server
+          _ => ctx.stop(),
+        }
+        fut::ready(())
+      })
+      .wait(ctx);
+  }
+  fn stopping(&mut self, _: &mut Self::Context) -> Running {
+    // notify chat server
+    self
+      .apub_data
+      .chat_server()
+      .do_send(Disconnect { id: self.id });
+    Running::Stop
+  }
 }
 
-async fn handle_messages(
-  mut stream: MessageStream,
-  client_ip: IpAddr,
-  mut session: Session,
-  connection_id: ConnectionId,
-  alive: Arc<Mutex<Instant>>,
-  rate_limiter: web::Data<RateLimitCell>,
-  context: ContextData<LemmyContext>,
-) -> Result<(), LemmyError> {
-  while let Some(Ok(msg)) = stream.next().await {
+/// Handle messages from chat server, we simply send it to peer websocket
+impl Handler<WsMessage> for WsChatSession {
+  type Result = ();
+
+  fn handle(&mut self, msg: WsMessage, ctx: &mut Self::Context) {
+    ctx.text(msg.0);
+  }
+}
+
+/// WebSocket message handler
+impl StreamHandler<Result<ws::Message, ws::ProtocolError>> for WsChatSession {
+  fn handle(&mut self, msg: Result<ws::Message, ws::ProtocolError>, ctx: &mut Self::Context) {
+    let msg = match msg {
+      Err(_) => {
+        ctx.stop();
+        return;
+      }
+      Ok(msg) => msg,
+    };
+
+    debug!("WEBSOCKET MESSAGE: {msg:?}"); // TODO remove this
     match msg {
-      ws::Message::Ping(bytes) => {
-        if session.pong(&bytes).await.is_err() {
-          break;
-        }
+      ws::Message::Ping(msg) => {
+        self.hb = Instant::now();
+        ctx.pong(&msg);
       }
       ws::Message::Pong(_) => {
-        let mut lock = alive
-          .lock()
-          .expect("Failed to acquire websocket heartbeat alive lock");
-        *lock = Instant::now();
+        self.hb = Instant::now();
       }
       ws::Message::Text(text) => {
-        let msg = text.trim().to_string();
-        let executed = parse_json_message(
-          msg,
-          client_ip.clone(),
-          connection_id,
-          rate_limiter.get_ref(),
-          context.reset_request_count(),
-        )
-        .await;
-
-        let res = executed.unwrap_or_else(|e| {
-          error!("Error during message handling {}", e);
-          e.to_json()
-            .unwrap_or_else(|_| String::from(r#"{"error":"failed to serialize json"}"#))
+        // TODO No way around these clones?
```

Comment:
The other thing. I don't know if there's any way around doing these clones.

Reply:
I think these are mostly pointers or a struct consisting of pointers, so not much data to clone. You can see with std::mem::size_of how big they are. For big data it makes sense to wrap it in a pointer type like `Arc` to avoid cloning, but for small types such as context that would likely be worse for performance due to extra pointer redirection (would have to follow pointer to get context struct for every method call).